plutus-pab/src/Cardano/Wallet/Server.hs

```diff
@@ -38,34 +38,46 @@ import           Servant                         (Application, Handler (Handler)
                                                   err400, err500, hoistServer, serve, (:<|>) ((:<|>)))
 import           Servant.Client                  (BaseUrl (baseUrlPort), ClientEnv, ClientError, mkClientEnv)
 
-import           Plutus.PAB.Monitoring           (runLogEffects)
+import           Plutus.PAB.Monitoring           (convertLog, handleLogMsgTrace, runLogEffects)
 import           Wallet.Effects                  (ChainIndexEffect, NodeClientEffect, WalletEffect, ownOutputs,
                                                   ownPubKey, startWatching, submitTxn, updatePaymentWithChange,
                                                   walletSlot)
 import           Wallet.Emulator.Error           (WalletAPIError)
 import           Wallet.Emulator.Wallet          (WalletState, emptyWalletState)
 import qualified Wallet.Emulator.Wallet          as Wallet
 
-type AppEffects m = '[WalletEffect, NodeClientEffect, ChainIndexEffect, State WalletState, LogMsg Text, Error WalletAPIError, Error ClientError, Error ServerError, m]
-
+type AppEffects m = '[WalletEffect
+                     , NodeClientEffect
+                     , ChainIndexEffect
+                     , State WalletState
+                     , LogMsg Text
+                     , Error WalletAPIError
+                     , Error ClientError
+                     , Error ServerError
+                     , m]
 runAppEffects ::
-    ( MonadIO m
-    )
-    => ClientEnv
+     MonadIO m
+    => Trace m WalletMsg
+    -> ClientEnv
     -> ClientEnv
     -> WalletState
     -> Eff (AppEffects m) a
     -> m (Either ServerError (a, WalletState))
-runAppEffects nodeClientEnv chainIndexEnv walletState action =
-    runM
-            $ runError
-            $ flip handleError (\e -> throwError $ err500 { errBody = Char8.pack (show e) })
-            $ flip handleError (throwError . fromWalletAPIError)
-            $ runStderrLog
-            $ runState walletState
-            $ ChainIndexClient.handleChainIndexClient chainIndexEnv
-            $ NodeClient.handleNodeClientClient nodeClientEnv
-            $ Wallet.handleWallet action
+runAppEffects trace nodeClientEnv chainIndexEnv walletState action =
+    Wallet.handleWallet action
+    & NodeClient.handleNodeClientClient nodeClientEnv
+    & ChainIndexClient.handleChainIndexClient chainIndexEnv
+    & runState walletState
+    & handleLogMsgTrace (toWalletMsg trace)
+    & handleWalletApiErrors
+    & handleClientErrors
+    & runError
+    & runM
+        where
+            handleWalletApiErrors = flip handleError (throwError . fromWalletAPIError)
```

Comment:
I would say "give these type signatures"... but I know what we're dealing with here :laughing: